### PR TITLE
TWAP notes/fixes from code review

### PR DESCRIPTION
**Pull request by ValarDragon**
## What is the purpose of the change

Notes / fixes from the live TWAP code review

Two important bugs found during code review! 

* Time.Now() bug in query that was getting vendored (fixed in this PR)
  * Thankfully wouldn't have caused mainnet issue, since exec time always greater than block time, and we've well defined returning an error in this case
* Bug in pruning logic (not yet fixed)

## Testing and Verifying

Mostly comment + TODO adds

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/twap/logic.go [hunk 1]
```diff
@@ -102,6 +102,8 @@ func (k Keeper) updateRecords(ctx sdk.Context, poolId uint64) error {
 	if err != nil {
 		return err
 	}
+	// TODO: Add a safety assert, that # of records is as we expect, given # of denoms in the pool
+	// namely, that for `k` denoms in pool, there should be k * (k - 1) / 2 records
```
Comment (ValarDragon) [inline x/twap/logic.go:106]: Extract into issue

Comment (czarcas7ic) [reply x/twap/logic.go:106]: https://github.com/osmosis-labs/osmosis/issues/2644

File: x/twap/store.go [hunk 2]
```diff
@@ -75,18 +75,27 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
 // - Suppose pruning param -48 hour
-// - Suppose swaps at -50 hour, -1hour
-// - A prune would leave us with only one record at -1 hour, and we are not able to get twaps from the
+// - Suppose there are three records, at -51 hour, -50 hour, -1hour
+// - A prune of everything older than 48 hours,
+// - we would be leave with with only one record at -1 hour, and we are not able to get twaps from the
 // [-48 hour, -1 hour] time range.
+// So in order to have correct behavior, for the desired guarantee,
+// we keep the newest record, that is older than the pruning time.
+// So then we keep -50 hour, and -1hour
```
Comment (czarcas7ic) [inline x/twap/store.go:84]:
```suggestion
// - Suppose there are three records at: -51 hour, -50 hour, and -1hour
// If we were to prune everything older than 48 hours,
// we would be left with with only one record at -1 hour, and we wouldn't be able to
// get twaps from the [-48 hour, -1 hour] time range.
// So, in order to have correct behavior for the desired guarantee,
// we keep the newest record that is older than the pruning time.
// This is why we would keep the -50 hour and -1hour twaps despite a 48hr pruning period
```

File: x/twap/store.go [hunk 3]
```diff
@@ -75,18 +75,27 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
 	// Due to how it is indexed, we will only iterate times starting from
 	// lastKeptTime exclusively down to the oldest record.
-	iter := store.ReverseIterator([]byte(types.HistoricalTWAPTimeIndexPrefix), types.FormatHistoricalTimeIndexTWAPKey(lastKeptTime, 0, "", ""))
+	iter := store.ReverseIterator(
+		[]byte(types.HistoricalTWAPTimeIndexPrefix),
+		types.FormatHistoricalTimeIndexTWAPKey(lastKeptTime, 0, "", ""))
 	defer iter.Close()
 
+	// TODO: REWRITE this, its bugged for multi-asset (3 or more) pools.
+	// needs the key to be {pool ID, asset 0, asset 1}
```
Comment (czarcas7ic) [inline x/twap/store.go:98]:
```suggestion
```
https://github.com/osmosis-labs/osmosis/pull/2642

File: x/twap/store.go [hunk 4]
```diff
@@ -177,6 +186,8 @@ func (k Keeper) getRecordAtOrBeforeTime(ctx sdk.Context, poolId uint64, t time.T
 	// Note that we cannot get any time entries from t + 1ns, as the key would be `prefix|t+1ns`
 	// and the end for a reverse iterator is exclusive. Thus the largest key that can be returned
 	// begins with a prefix of `prefix|t`
+	// TODO: Consider separator tricks, to not have the + 1 ns
+	// TODO: Change key prefix indexing, to include assets, and not re-order assets at end.
```
Comment (czarcas7ic) [inline x/twap/store.go:190]:
```suggestion
```
https://github.com/osmosis-labs/osmosis/pull/2641

File: CHANGELOG.md [hunk 5]
```diff
@@ -40,7 +40,17 @@ All notable changes to this project will be documented in this file.
...
 
-## Unreleased
+## v12.0.0
+
+This release includes several cosmwasm-developer and appchain-ecosystem affecting upgrades:
+
+* TWAP - Time weighted average prices for all AMM pools
+* Cosmwasm contract developers
+  * Enabling select queries for cosmwasm contracts
+  *
```
Comment (czarcas7ic) [inline CHANGELOG.md:50]:
```suggestion
```

**Review by czarcas7ic — APPROVED** [review]
👍

File: x/twap/store.go [hunk 6]
```diff
@@ -75,18 +75,27 @@ func (k Keeper) storeHistoricalTWAP(ctx sdk.Context, twap types.TwapRecord) {
...
-// - Suppose swaps at -50 hour, -1hour
-// - A prune would leave us with only one record at -1 hour, and we are not able to get twaps from the
+// - Suppose there are three records, at -51 hour, -50 hour, -1hour
+// - A prune of everything older than 48 hours,
+// - we would be leave with with only one record at -1 hour, and we are not able to get twaps from the
 // [-48 hour, -1 hour] time range.
+// So in order to have correct behavior, for the desired guarantee,
+// we keep the newest record, that is older than the pruning time.
+// So then we keep -50 hour, and -1hour
+// TODO: RENAME THIS FUNCTION TO BE MORE ACCURATE
```
Comment (mattverse) [inline x/twap/store.go:85]: I propose `keepNewestRecordAndPrune`

**Review by mattverse — APPROVED** [review]
Straight foward lgtm

File: x/twap/store.go (hunk 6 above)
Comment (ValarDragon) [reply x/twap/store.go:85]: Perfect name, nice

**Pull request merged**